ROADMAP.md

```diff
@@ -0,0 +1,51 @@
+# js-libp2p roadmap Q4’22/Q1’23
+
+## 🛣️ Milestones
+### 2022
+
+#### Early Q4 (October)
+- A.1📺 Universal Browser Connectivity - [**WebTransport**](https://github.com/libp2p/js-libp2p-webtransport/issues/1)
+
+#### End of Q4 (December)
+- A.2 📺 Universal Browser Connectivity - [**WebRTC**](https://github.com/little-bear-labs/js-libp2p-webrtc/pull/4) (Browser to Server)
+- B.1 🥊 Decentralized Hole Punching - [**AutoNat**](https://github.com/libp2p/js-libp2p/issues/1005)
+- B. 2 🥊 Decentralized Hole Punching - [**Circuit Relay v2**](https://github.com/libp2p/js-libp2p/issues/1029)
+
+### 2023
+
+#### End of Q1 (March)
+- B. 3 🥊 Decentralized Hole Punching - **Hole Punching**
+	- 🎉 Estimated Project Completion
+- C. 🧪 Future-proof Testing
+
+
+# Roadmap Appendix
+## A. 📺 Universal Browser Connectivity
+<!--- TODO: Link to GitHub Epic -->
+
+**Why**: A huge part of “the Web” is happening inside the browser. As a universal p2p networking stack, libp2p needs to be able to offer solutions for browser users.
+
+**Goal**: js-libp2p supports both WebTransport and (libp2p-) WebRTC protocols, enabled by default. This allows connections between browsers and public nodes, and eventually between browsers and non-public nodes and in between browser nodes.
+
+1. [WebTransport](https://github.com/libp2p/js-libp2p-webtransport/issues/1): Implementation of WebTransport in js-libp2p
+2. [WebRTC](https://github.com/little-bear-labs/js-libp2p-webrtc/pull/4): while browser to public node is getting close to finalized, there’ll be a push to make the other combinations work as well
+
+## B. 🥊 Decentralized Hole Punching
+<!--- TODO: Link to GitHub Epic -->
+**Why**:  P2P networks can have a combination of both public and private nodes. While private nodes can dial nodes on the public Internet, they are unreachable from the outside as they are behind a NAT or a firewall. We need a mechanism to dial them. A [previous DHT crawl found that almost 63%](https://github.com/libp2p/specs/blob/master/ROADMAP.md#-hole-punching-on-tcp-and-quic) of the network was undialable.  This project aims to implement Decentralized Hole Punching in js-libp2p and bring it to parity with the go and rust implementations.
+
+**Goal**:
+1. [AutoNat](https://github.com/libp2p/js-libp2p/issues/1005) - Determine whether a node is public or private (located behind a firewall or a NAT.)
+2. [Circuit Relay v2](https://github.com/libp2p/js-libp2p/issues/1029) - Connect to,  request reservations, and establish a secure relay connection through discovered public relay node.
+3. Hole Punching - Use [DCUtR](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md) to synchronize hole punching
+
+Dependencies: B.3 Hole punching is more reliable with QUIC as a transport and thus requires finding a way to use QUIC in node.js
```

Comment:
What's the data on this?  I didn't see it in https://www.notion.so/pl-strflt/NAT-Hole-punching-Success-Rate-2022-09-29-Data-Analysis-8e72705ca3cc49ab983bc5e8792e3e98

Unless we have clear data and the data is super compelling, I wouldn't discuss this here.  We can discuss in future the merits of improving hole-punching by some percentage with QUIC once we have all the other pieces in place.  

If we do believe QUIC is paramount for this effort then adding a Quic transport should be its own milestone.

Reply:
> What's the data on this?

_I can only comment on UDP (thus QUIC) vs TCP._

There is general industry consensus that hole punching with UDP is more successful than hole punching with TCP. Our first project flare testing round confirmed this.

The graph below shows the same in the recent measurements from @dennis-tra.

https://www.notion.so/pl-strflt/NAT-Hole-punching-Success-Rate-2022-09-29-Data-Analysis-8e72705ca3cc49ab983bc5e8792e3e98#5b76991da8d24736abd486aa93e85a97